### ISSUE TITLE
Visual glitches main window (focus problem?)

### ISSUE THREAD

**Issue by mihaiile** [keepassxreboot/keepassxc#304]
I have KeepassXC set up to minimise to tray at start-up and I have noticed that when I open the main window for the first time after launching the application using "Toggle Main Window" option, the window appears but it flashes all around it (the window shadow always flashes and it seems to have something like a 1px white border around it).
This disappears as soon as I focus the window by clicking on it with the mouse and after that minimising/bringing it back no longer has this weird behaviour until I start the application again.

## Expected Behavior

No visual glitches around main window when showing it for the first time after application launching

## Current Behavior

Visual glitches are present around main window

## Steps to Reproduce (for bugs)

1. In application settings set hide to tray when minimised and minimise at startup
2. Close application and launch it again
3. From tray choose "Toggle window"
4. Notice visual glitches around main window (window shadow "shaking") which disappear when clicking with the mouse on the window (bringing it to focus)
5. Minimise window and from tray choose "Toggle window"
6. Notice visual glitches are no longer visible

## Your Environment

* KeePassXC version/commit used: 2.1.1 (app image and deb file); 2.1.2 branch (compiled locally)
* Qt version (e.g. Qt 5.3): qt 5.6.1 (and version from appimage also)
* Operating System and version: Ubuntu 16.10

I have attached a small video to show but unfortunately the video does not capture the shacking of the shadow around the window but you can notice that in the top corners there are some lines around the window and that the shadow is very strong around it which all disappear when clicking on the window.
[out.ogv.zip](https://github.com/keepassxreboot/keepassxc/files/773624/out.ogv.zip)

**Comment by phoerious** [comment]
I was able to reproduce that problem on an Ubuntu 16.04 machine. I'm not getting this problem at home in my 16.04 VM or on my KDE machines.
I assume this must be a compositing problem in combination with your graphics driver. What graphics card/driver are you using?

I built the current 2.1.2 release branch and the problem seems to be a little better with that version (meaning it only appears once at application startup, but not when I continuously toggle the window). I'm not sure, though, if we can do anything about it. I can try some fancy things with window states, but basically, this problem is beyond your reach.

**Comment by mihaiile** [comment]
Desktop PC uses nvidia drivers, but also happens on a laptop with intel card/default ubuntu drivers.
I would assume that might be something with the focus states since when you launch the application you see the window trying to appear just to see a ui glitch of it minimising which is also not very nice.

Could this issue, my above statement and maybe the menu showing up/not showing in ubuntu problem (that I am not sure is fixed for good) be avoided by having a "Close to tray" instead of "minimise to tray"?

I am using skype alpha version and I think they also use qt and they are using "close to tray" instead of minimise, so you actually close the main window, the icon in the dock goes away and you are left just with the tray icon. If you minimise, it just minimises as normal.

**Comment by phoerious** [comment]
There is no minimize to tray. Minimizing to tray is effectively closing the window. That's what's causing the trouble in the first place.
I also noticed that my previous fix for the bloody global menu on Ubuntu doesn't work with appmenu-qt5, because the isNativeMenuBar() member of the menu bar doesn't realize that it actually is a global menu.

I managed to work around the compositing glitch by not showing the window at all instead of showing it shortly and then minimizing/closing it immediately. I prepared a fix, but need to test it on other platforms.

**Comment by mihaiile** [comment]
Currently you have to minimise the main window so that it will hide from dock. "Toggle window" will restore or minimise the window again.
I was saying that maybe it would be better to actually leave the minimise button as expected and let the application minimise and instead make the X button to actually close the window (removing it from dock) but still leaving the application in tray.

**Comment by phoerious** [comment]
That is a setting. You can enable "Hide window to system tray instead of app exit" without enabling "Hide window to system tray when minimized".

**Comment by mihaiile** [comment]
Yes, but visually what I see is that when I click X on the window, it minimises to the dock and after that the dock icon disappears. If it was actually closing, the window would just disappear instead of the minimising animation&icon-hiding.

Also when application is starting I can see that a window is being created and minimised, just some UI glitches since it is happening very fast but noticeable still.

I think that the application should not even start the main window if we have "Minimise window at app startup" enabled. I mean what is the point in launching a window and minimising it automatically, it just creates this ugly UI glitches of something wanting to appear just to be minimised after.

**Comment by phoerious** [comment]
The minimizing before closing is actually done to prevent such glitches and create a better illusion of "minimizing to tray". These problems on Ubuntu appear exactly when the window is simply shown/hidden without any prior minimization.
But as I said: i have a working fix for everything, I only need to test it on other platforms before we can merge it.

**Comment by phoerious** [comment]
You can try out the `hotfix/304-minimize-tray-decoration-glitches` branch if you like.

**Comment by mihaiile** [comment]
Just tested, it is better now, thanks :+1:

**Comment by phoerious** [comment]
Better or good?

**Comment by mihaiile** [comment]
It is good, it solves the problem and also starting the application does not show any glitches

**Comment by phoerious** [comment]
Fixed in #307 and merged into the release branch.
